### PR TITLE
feat(elm): display answer evaluation notification

### PR DESCRIPTION
**Pull request by jtkpiotr**
**Summary**

**Related issues**

**Test plan**

### CODE REVIEW

File: aion/web/elm/src/Update.elm [hunk 1]
```diff
@@ -89,12 +91,15 @@ update msg model =
                 push_ =
                     Phoenix.Push.init "new:answer" ("rooms:" ++ (toString model.roomId))
                         |> Phoenix.Push.withPayload payload
+                        |> Phoenix.Push.onError (\v -> WrongAnswer)
```
Comment (pmrukot) [inline aion/web/elm/src/Update.elm]: Why do we invoke `WrongAnswer` on push? `ReceiveFeedback` is much better choice and it is already prepared to use:
https://github.com/pmrukot/aion/blob/623b9549e929f16f28bdca03d55e32cf83967bfe/aion/web/elm/src/Update.elm#L76

Also it does not feel scalable since we would like to have different types of notifications, not only `onError`. Currently we have three types of feedback message + we would like to have also when user joins the room, leaves the room, scores the point etc.

Details about feedback are in this PR:
https://github.com/pmrukot/aion/pull/48/files#diff-1d980e9fadf285c6fee743d51b4c8e6dR60

File: aion/web/elm/src/Room/Models.elm [hunk 2]
```diff
@@ -30,7 +30,8 @@ type alias UserList =
 
 
 type alias UserGameData =
-    { currentAnswer : String
+    {
+      currentAnswer : String
     }
```
Comment (pmrukot) [inline aion/web/elm/src/Room/Models.elm]: Please follow `elm-format`: https://github.com/avh4/elm-format
You can integrate it with your IDE and add watcher for files.
https://github.com/avh4/elm-format#editor-integration

File: aion/web/elm/src/Update.elm (hunk 1 above)
Comment (pmrukot) [reply aion/web/elm/src/Update.elm]: Something like
```elm
case answerFeedback.feedback do
  "Correct" -> correctAnswerToast (model ! [])
  "Close one!" -> closeAnswerToast (model ! [])
```
etc.

**Review by pmrukot — CHANGES_REQUESTED** [review]

File: aion/web/channels/subject_channel.ex [hunk 3]
```diff
@@ -21,16 +21,18 @@ defmodule Aion.SubjectChannel do
...
-    end
-
-    {:noreply, socket}
+     case evaluation do
+       1.0 ->
+         send_user_list(socket, room_id)
+         send_new_question(socket, room_id)
+         {:noreply, socket}
+       _ ->
+         {:reply, :error, socket}
```
Comment (mrapacz) [inline aion/web/channels/subject_channel.ex]: Error suggests that there is a problem with the program's flow, which is not true, because even if the user answers incorrectly, it's still proper from the program's point of view. I would save atom's like :error for situations when something unexpected appears. (something like :wronganswer would do much better imo)

File: aion/web/elm/src/Update.elm (hunk 1 above)
Comment (mrapacz) [reply aion/web/elm/src/Update.elm]: I have to agree with the point about scalability etc. In general the flow of the answerFeedback is totally the same, regardless of whether it is correct, almost correct or incorrect, therefore I would stick to sending the valuable information in one string and then displaying it differently depending on its content.

I believe we can do it as @pmrukot suggests or we could even send `evaluation` from elixir to elm and then assign certain strings on frontend side.

File: aion/web/channels/subject_channel.ex [hunk 4]
```diff
@@ -26,8 +26,8 @@ defmodule Aion.SubjectChannel do
     send_feedback socket, evaluation
 
     if evaluation == 1.0 do
-      send_user_list(socket, room_id)
-      send_new_question(socket, room_id)
```
Comment (mrapacz) [inline aion/web/channels/subject_channel.ex]: Fix indentation to two spaces

File: aion/web/elm/src/Room/Notifications.elm [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+
+incorrectAnswerToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+incorrectAnswerToast = addToast (Toasty.Defaults.Error "Error!" "Wrong Answer")
+
+closeAnswerToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+closeAnswerToast = addToast (Toasty.Defaults.Warning "Close one!" "Your Answer is Almost Correct")
+
+correctAnswerToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+correctAnswerToast = addToast (Toasty.Defaults.Success "Good Answer!" "You Answer is Correct")
```
Comment (mrapacz) [inline aion/web/elm/src/Room/Notifications.elm]: Just a typo "You" instead of "Your"

File: aion/web/elm/src/Update.elm [hunk 6]
```diff
@@ -89,6 +95,7 @@ update msg model =
                 push_ =
                     Phoenix.Push.init "new:answer" ("rooms:" ++ (toString model.roomId))
                         |> Phoenix.Push.withPayload payload
+                        |> Phoenix.Push.onOk (\v -> ReceiveAnswerFeedback v)
```
Comment (mrapacz) [inline aion/web/elm/src/Update.elm]: What is `v` here?

File: aion/web/elm/src/Update.elm [hunk 7]
```diff
@@ -70,10 +72,14 @@ update msg model =
...
                 Ok answerFeedback ->
                     let
-                        x =
-                            Debug.log "feedback" answerFeedback.feedback
+                        answerToast =
+                            case answerFeedback.feedback of
+                                "incorrect" -> incorrectAnswerToast
+                                "close" -> closeAnswerToast
+                                "correct" -> correctAnswerToast
+                                _ -> (\x -> x)
```
Comment (mrapacz) [inline aion/web/elm/src/Update.elm]: We will probably remodel the evaluation part in the future, so we could just go with Debug.crash here or at least log an error here.

**Review by mrapacz — CHANGES_REQUESTED** [review]
Looks much better after changes, only minor fixes left right now I guess

**Comment by pmrukot** [comment]
I've tested the branch and it looked really nice, just need to rebase, run elm-format and correct minor issues above and we are good to go

File: aion/config/dev.exs [hunk 8]
```diff
@@ -36,8 +36,8 @@ config :phoenix, :stacktrace_depth, 20
 # Configure your database
```
Comment (mrapacz) [inline aion/config/dev.exs]: No need to commit changes to this file actually as they are generated by docker and are not part of the PR

File: aion/web/elm/src/Room/Notifications.elm [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+addToast : Toasty.Defaults.Toast -> ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+addToast toast ( model, cmd ) =
+    Toasty.addToast myConfig ToastyMsg toast ( model, cmd )
+
+
+incorrectAnswerToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+incorrectAnswerToast = addToast (Toasty.Defaults.Error "Error!" "Wrong Answer")
+
+closeAnswerToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+closeAnswerToast = addToast (Toasty.Defaults.Warning "Close one!" "Your Answer is Almost Correct")
```
Comment (mrapacz) [inline aion/web/elm/src/Room/Notifications.elm]: Maybe you could just import the Toasty.Defaults and Toasty itself exposing what you want to use and get rid of the module names, it would make the code more legible imo

Comment (mrapacz) [reply aion/web/elm/src/Room/Notifications.elm]: But it's arguable, you can do it the way you want I think

File: .gitignore [hunk 10]
```diff
@@ -2,4 +2,6 @@
 .DS_Store
 /postgres-data
 *.pyc
+aion/web/static/assets/images/
+.gitignore
```
Comment (mrapacz) [inline .gitignore]: What's the reason for putting .gitignore here?

File: aion/web/elm/src/Room/Notifications.elm (hunk 9 above)
Comment (jtkpiotr) [reply aion/web/elm/src/Room/Notifications.elm]: Toasty.Defaults does not expose Error, Warning or Success. I will leave it as it is ;)

**Comment by pmrukot** [comment]
I would love to merge it but code still needs elm format since it changes:
https://github.com/pmrukot/aion/pull/59/files#diff-db0c365650eff9af07bd5a9aee73dba4R33

Its easy to integrate with IntelliJ: 
https://github.com/avh4/elm-format
https://github.com/avh4/elm-format#jetbrains-installation

Other than that, edit the PR name to
`feat(elm): Add answer evaluation notifications`
or something along those lines. And remember to do squash&merge instead of regular merge.

**Review by pmrukot — APPROVED** [review]

**Comment by pmrukot** [comment]
@mrapacz Have a look now 👋

**Review by mrapacz — APPROVED** [review]
